sql/src/main/java/io/crate/operation/scalar/conditional/IfFunction.java

```diff
@@ -119,9 +120,13 @@ private static FunctionInfo createInfo(List<DataType> dataTypes) {
         DataType returnType = valueType;
         if (dataTypes.size() == 3) {
             returnType = dataTypes.get(2);
-            if ((returnType.id() != valueType.id()) && (valueType != DataTypes.UNDEFINED) && (returnType != DataTypes.UNDEFINED)) {
+            // allow different types only in case of null or primitive numeric types that can be converted
```

Comment:
can't you use isConvertableTo() method on dataType instead of specially handling the numeric types?